pkg/controllers/nodetemplate/infrastructure.go

```diff
@@ -0,0 +1,153 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodetemplate
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/multierr"
+	"k8s.io/client-go/util/workqueue"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/metrics"
+	awssettings "github.com/aws/karpenter/pkg/apis/config/settings"
+	"github.com/aws/karpenter/pkg/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/controllers/providers"
+)
+
+type Infrastructure struct {
+	kubeClient          client.Client
+	sqsProvider         *providers.SQS
+	eventBridgeProvider *providers.EventBridge
+
+	lastInfrastructureReconcile time.Time // Keeps track of the last reconcile time for infra, so we don't keep calling APIs
+}
+
+// Reconcile reconciles the infrastructure based on whether interruption handling is enabled and deletes
+// the infrastructure by ref-counting when the last AWSNodeTemplate is removed
+func (i *Infrastructure) Reconcile(ctx context.Context, nodeTemplate *v1alpha1.AWSNodeTemplate) (reconcile.Result, error) {
+	if awssettings.FromContext(ctx).EnableInterruptionHandling {
+		list := &v1alpha1.AWSNodeTemplateList{}
+		if err := i.kubeClient.List(ctx, list); err != nil {
+			return reconcile.Result{}, err
+		}
+		if !nodeTemplate.DeletionTimestamp.IsZero() && len(list.Items) == 1 {
+			if err := i.DeleteInfrastructure(ctx); err != nil {
+				return reconcile.Result{}, err
+			}
+			i.lastInfrastructureReconcile = time.Time{}
+			return reconcile.Result{}, nil
```

Comment:
Should this logic be handled in a finalizer? You have one elsewhere.

Reply:
I don't know what you mean here. This entire sub-reconciler handles the infrastructure so that it decouples that from the finalizer creation/deletion logic